proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -24,77 +24,148 @@ vector/matrix operations described in [0029].
+* works well with existing HLSL APIs
 * is expected to work well with future HLSL APIs in the same problem space
 * can be implemented reasonably in DXC and cleanly in clang
 
-This design builds on the "long vectors" feature described in [0026].
-
 [0026]: 0026-hlsl-long-vector-type.md
 
 ## Proposed solution
 
-First strawman:
+This API will be implemented using HLSL code.  The exact mechanism for getting
+this code into a developer's shader is TBD, but implementations have a few
+possible options, including:
+
+* Developers must explicitly #include a header file
+* The compiler force-includes the header file
+* The compiler force-includes a precompiled version of the header file
+
+The header-implementation accesses the DXIL operations described in [0029] by
+calling low-level builtins. These builtins should be considered implementation
+details and users should not call them directly. However, since they are a part
+of the implementation plan for the first implementation of this API they are
+described [below](#builtins).
+
+Since this API is currently only supported by DirectX, all the new types /
+methods described in it are placed in the `dx` namespace. Within this namespace,
+a `linalg` namespace is also added to group together types and methods related
+to linear algebra.
+
+Throughout this API, template parameters are used to store values that must be
+known at compile time, while member variables or function arguments are used to
+store values that may only be determined at runtime.
+
+This API defines the following supporting types:
+
+* `struct dx::linalg::MatrixRef`
+  * Reference to a matrix stored in a ByteAddressBuffer.   
+* `struct dx::linalg::RWMatrixRef`
+  * Reference to a matrix stored in a RWByteAddressBuffer.
+* `struct dx::linalg::VectorRef`
+  * Reference to a vector stored in a ByteAddressBuffer.
+* `struct dx::linalg::RWVectorRef`
+  * Reference to a vector stored in a RWByteAddressBuffer.
+* `struct dx::linalg::Vector`
+  * Wrapper around a vector, allowing the elements of the vector to be
+    reinterpreted in various ways.
+* `enum dx::linalg::DataType`
+  * Enum describing various data types that can be used to applied to matrices
+    and vectors.
+* `enum dx::linalg::MatrixLayout`
+  * Enum describing the possible layouts for a matrix in memory.
+
+This API defines the following functions:
+
+* `dx::linalg::Mul`
+  * Multiply a matrix in memory by a vector parameter.
+* `dx::linalg::MulAdd`
+  * Multiply a matrix in memory by a vector parameter, and add a vector from
+    memory.
+* `dx::linalg::OuterProductAccumulate`
+  * Compute the outer product of two vectors and accumulate the result matrix
+    atomically-elementwise in memory.
+* `dx::linalg::VectorAccumulate`
+  * Accumulate elements of a vector atomically-elementwise to corresponding
+    elements in memory.
+* `dx::linalg::InterpretedVector`
+  * Convenience function to construct a `Vector` inline while inferring various
+    template parameters.
```

Comment:
Pending #441 to see if this is really necessary as-is, but if that issue indicates that we do really need the expressiveness then this looks good to me.